### PR TITLE
[YUNIKORN-961] Expose build and version information in a REST call

### PR DESCRIPTION
**Pull request by steinsgateted**
### What is this PR for?
in k8shim:
save go build version and build date, and pass to core via scheduler-interface.

in core:
save k8shim's information and the time the scheduler was started.
then the Rest call pushes all information out.

in scheduler-interface:
the message RegisterResourceManagerRequest is extended with the build data.
add a map<string, string> buildInfo = 4 to the message.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/projects/YUNIKORN/issues/YUNIKORN-961

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Comment by steinsgateted** [comment]
It has not been completed, execute make test in k8sshim on the local side, there will be this error:

`fatal error: sync: RUnlock of unlocked RWMutex`

detailed:
```
2022-02-09T00:02:52.602+0800    INFO    entrypoint/entrypoint.go:46     ServiceContext start all services
2022-02-09T00:02:52.606+0800    INFO    entrypoint/entrypoint.go:93     ServiceContext start scheduling services
2022-02-09T00:02:52.607+0800    INFO    entrypoint/entrypoint.go:110    creating InternalMetricsHistory
2022-02-09T00:02:52.607+0800    INFO    entrypoint/entrypoint.go:117    ServiceContext start web application service
2022-02-09T00:02:52.610+0800    INFO    webservice/webservice.go:71     web-app started {"port": 9080}
2022-02-09T00:02:52.610+0800    INFO    log/logger.go:89        scheduler configuration, pretty print   {"configs": "{\n \"schedulerName\": \"yunikorn\",\n \"clusterId\": \"my-kube-cluster\",\n \"clusterVersion\": \"0.1\",\n \"policyGroup\": \"queues\",\n \"schedulingIntervalSecond\": 1000000000,\n \"absoluteKubeConfigFilePath\": \"\",\n \"loggingLevel\": 0,\n \"logEncoding\": \"console\",\n \"logFilePath\": \"\",\n \"volumeBindTimeout\": 10000000000,\n \"testMode\": true,\n \"eventChannelCapacity\": 1048576,\n \"dispatchTimeout\": 300000000000,\n \"kubeQPS\": 1000,\n \"kubeBurst\": 1000,\n \"predicates\": \"\",\n \"operatorPlugins\": \"general,yunikorn-app\",\n \"enableConfigHotRefresh\": false,\n \"disableGangScheduling\": false,\n \"userLabelKey\": \"yunikorn.apache.org/username\"\n}"}
2022-02-09T00:02:52.610+0800    INFO    appmgmt/appmgmt.go:50   Initializing new AppMgmt service
2022-02-09T00:02:52.611+0800    INFO    dispatcher/dispatcher.go:80     Init dispatcher {"EventChannelCapacity": 1048576, "AsyncDispatchLimit": 104857, "DispatchTimeoutInSeconds": 300}
2022-02-09T00:02:52.611+0800    WARN    appmgmt/appmgmt.go:133  App manager is not registered   {"app manager name": "yunikorn-app"}
2022-02-09T00:02:52.611+0800    INFO    dispatcher/dispatcher.go:183    starting the dispatcher
2022-02-09T00:02:52.612+0800    INFO    cache/placeholder_manager.go:144        starting the PlaceholderManager
2022-02-09T00:02:52.612+0800    INFO    shim/scheduler_mock_test.go:119 waiting for scheduler state     {"expected": "Running", "actual": "New"}
fatal error: sync: RUnlock of unlocked RWMutex

goroutine 76 [running]:
```

**Comment by yangwwei** [comment]
hi @steinsgateted there are some lint errors, could you please fix them? you can verify this locally with `make lint` before submitting the patch. thanks!

**Comment by steinsgateted** [comment]
Hi @yangwwei,
Thanks for the comment.
I run make lint with this message:
```
pkg/shim/scheduler.go:254:3: unknown field `BuildInfo` in struct literal (typecheck)
                BuildInfo:   buildInfoMap,
```
I guess it's because in scheduler-interface hasn't updated build info yet.

**Comment by yangwwei** [comment]
> Hi @yangwwei, Thanks for the comment. I run make lint with this message:
> 
> ```
> pkg/shim/scheduler.go:254:3: unknown field `BuildInfo` in struct literal (typecheck)
>                 BuildInfo:   buildInfoMap,
> ```
> 
> I guess it's because in scheduler-interface hasn't updated build info yet.

That's correct. Can you please update the dependency in this PR? You can do this by go get command. Thanks!

File: pkg/cmd/schedulerplugin/main.go [hunk 1]
```diff
@@ -40,6 +40,9 @@ func main() {
 	pluginconf.BuildVersion = version
 	pluginconf.BuildDate = date
 
+	conf.Version = version
+	conf.Date = date
+
```
Comment (wilfred-s) [inline pkg/cmd/schedulerplugin/main.go:45]: These should replace the plugin config BuildVersion and BuildDate, those should be removed here and from the pluginconf code

File: pkg/conf/schedulerconf.go [hunk 2]
```diff
@@ -51,6 +51,9 @@ var once sync.Once
 var configuration *SchedulerConf
 var factory = initConfigs
 
+var Version string
+var Date string
+
```
Comment (wilfred-s) [inline pkg/conf/schedulerconf.go:56]: Add a boolean flag to show that it is the plugin version or not. Set to true in the schedulerplugin/main.go file

File: pkg/shim/scheduler.go [hunk 3]
```diff
@@ -241,16 +242,23 @@ func (ss *KubernetesShim) doScheduling(e *fsm.Event) {
...
+
+	var buildInfoMap map[string]string
+	buildInfoMap["version"] = conf.Version
+	buildInfoMap["date"] = conf.Date
+
 	registerMessage := si.RegisterResourceManagerRequest{
 		RmID:        configuration.ClusterID,
 		Version:     configuration.ClusterVersion,
 		PolicyGroup: configuration.PolicyGroup,
+		BuildInfo:   buildInfoMap,
```
Comment (wilfred-s) [inline pkg/shim/scheduler.go:254]: This should be fixed as the SI reference has been updated already to include this

File: pkg/shim/scheduler.go [hunk 4]
```diff
@@ -241,16 +242,23 @@ func (ss *KubernetesShim) doScheduling(e *fsm.Event) {
...
 	}
 
 	log.Logger().Info("register RM to the scheduler",
 		zap.String("clusterID", configuration.ClusterID),
 		zap.String("clusterVersion", configuration.ClusterVersion),
-		zap.String("policyGroup", configuration.PolicyGroup))
+		zap.String("policyGroup", configuration.PolicyGroup),
+		zap.String("buildInfo", fmt.Sprintf("%v", buildInfoMap)))
```
Comment (wilfred-s) [inline pkg/shim/scheduler.go:261]: use the `zap.Any("buildInfo", buildInfoMap)` construct and let zap figure out how to log this.

**Review by wilfred-s — CHANGES_REQUESTED** [review]
We can finish these changes before we finalise the core. Commit will need to wait until after the core changes are done.

**Review by wilfred-s — CHANGES_REQUESTED** [review]
Sorry if I was not clear: I meant the other way around. Remove from `pluginconf` and keep in `conf`.
"/pkg/conf" is already imported by both the plugin main and the normal main and we do not want to pollute the non plugin code with the plugin as that could cause issues later.

**Comment by steinsgateted** [comment]
Sorry I made a mistake. Modified. Thanks.

**Review by wilfred-s — APPROVED** [review]
LGTM

**Pull request closed without merging**